drivers: watchdog: stm32: Support System Window Watchdog (WWDG)

gon1332 opened the pull request:
STM32 platform also support a WWDG, along with the IWDG.

The IWDG is clocked from the LSI, but the WWDG by the prescaled PCLK. Meaning that:
- IWDG cannot become disabled, even in Low Power States and reset. The WWDG becomes disabled in Low Power States and after a reset.
- IWDG allows for bigger and known timeouts, whereas the WWDG for smaller and dependent on what PCLK the MCU runs on.
- WWDG has a window (minimum and maximum timeout values) in which the watchdog must be fed.

This driver does not implement/focus on the window part of the WWDG. So, only the maximum timeout is taken into account. So, if you want to have a watchdog that can be disabled you can use this one over IWDG, with some little setbacks. However, it is not much of a work to also support the minimum timeout.

gon1332 committed:
dts: stm32: Add WWDG in STM32 dtsi files

Added System WWDG support for STM32 devices.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 committed:
soc: stm32: Add fixups for WWDG EWI and clock info

Extended the WWDG fixups to support early wake interrupt handling and
clock information.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 committed:
drivers: watchdog: stm32: Add implementation for WWDG

Added drivers for System Window Watchdog of STM32 platform.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 committed:
boards: arm: nucleo_{f401re, l496zg}: Enable WWDG

Added support for System Window Watchdog.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 committed:
samples: drivers: watchdog: Handle watchdog event

Also modified to timeout to 1000ms in order to support watchdogs like
WWDG with smaller timeouts.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 commented:
@galazari

erwango commented:
@gon1332, question before starting:
Seems you  didn't consider following series H7, G0, L1, WB and MP1.
Was there a reason for this choice?

gon1332 commented:
> @gon1332, question before starting:
> Seems you didn't consider following series H7, G0, L1, WB and MP1.
> Was there a reason for this choice?

Hm, the reason was mainly because I've seen that they do not support IWDG in Zephyr. So I didn't even searched for their docs in the first place. Also adapted the PR to master from our internal v1.14 version where these SOCs are not supported there. However I can also expand to them.

erwango commented:
> Hm, the reason was mainly because I've seen that they do not support IWDG in Zephyr. 

Ok, reason for that is purely because the were ported after IWDG introduction.

> So I didn't even searched for their docs in the first place. Also adapted the PR to master from our internal v1.14 version where these SOCs are not supported there. However I can also expand to them.

Ok, thanks to do that.

erwango commented on boards/arm/nucleo_f401re/Kconfig.defconfig:
```diff
@@ -44,4 +45,14 @@ config PWM_STM32_2
 
 endif # PWM
 
+if WATCHDOG
```
Please update board doc

gon1332 replied:
Done for both boards.

erwango commented on drivers/watchdog/wdt_wwdg_stm32.c:
```diff
@@ -0,0 +1,241 @@
+static int wwdg_stm32_init(struct device *dev)
+{
+	/* Enable the peripheral clock of WWDG */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_WWDG);
```
Please use clock API.

erwango commented on drivers/watchdog/wdt_wwdg_stm32.h:
```diff
@@ -0,0 +1,40 @@
+/* driver data */
+struct wwdg_stm32_data {
+	/* WWDG peripheral instance. */
+	WWDG_TypeDef *Instance;
```
Please set this field in a ROMable _config structure, along with clock settings.

erwango commented on drivers/watchdog/wdt_wwdg_stm32.c:
```diff
@@ -0,0 +1,241 @@
+
+	if (config->callback != NULL) {
+		data->callback = config->callback;
+		wwdg_stm32_irq_config(dev);
```
We generally call IRQ init function in device init function.
Calling it here, you'll call multiple times `IRQ_CONNECT` and `irq_enable`. 
I don't think this should be harmful, but this is not usual.

erwango commented on drivers/watchdog/wdt_wwdg_stm32.c:
```diff
@@ -0,0 +1,241 @@
+				       u32_t *prescaler,
+				       u32_t *counter)
+{
+	u32_t clock_freq = SystemCoreClock;
```
Shouldn't we rather consider the APB clock (PCLK1) here?
Should differ from `SystemCoreClock` by AHB and APB1 prescalers.

gon1332 commented on boards/arm/nucleo_f401re/Kconfig.defconfig:
```diff
@@ -44,4 +45,14 @@ config PWM_STM32_2
+config WWDG_STM32
+	default y
+
+config IWDG_STM32
```
Yes, because as you can [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/st_stm32/common/Kconfig.defconfig.series#L44) under the soc/ `IWDG_STM32` is enabled by default. In order to not break other things that use it I haven't change it.

gon1332 commented:
@erwango Could you recheck the changes on your comments?

The think that remains for sure is the porting of the other SOCs.

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -165,6 +166,15 @@
 		};
 
+		wwdg: watchdog@40002c00 {
+			compatible = "st,stm32-watchdog";
```
I don't think we should reuse same compatible as IWDG.
These are not similar IPs and hence can't share same 'compatible'.
Also, if used together (which is a valid use case), this will raise issue in fixup's handling.

gon1332 replied:
Done

erwango reviewed (changes requested):
@gon1332 , thanks for the update.
Main pending request would be to use WWDG specific compatible/binding

erwango commented on soc/arm/st_stm32/stm32f0/dts_fixup.h:
```diff
@@ -188,7 +188,13 @@
 #define DT_TIM_STM32_17_CLOCK_BUS	DT_ST_STM32_TIMERS_40014800_CLOCK_BUS
 
-#define DT_WDT_0_NAME                   DT_INST_0_ST_STM32_WATCHDOG_LABEL
+#define DT_WDT_0_NAME       DT_INST_0_ST_STM32_WATCHDOG_LABEL
```
Please fix indentation.
Apply everywhere

gon1332 replied:
I think it's good now.

erwango commented on drivers/watchdog/wdt_wwdg_stm32.c:
```diff
@@ -0,0 +1,272 @@
+
+static struct wwdg_stm32_config wwdg_stm32_dev_config = {
+	.pclken = {
+		.enr = DT_INST_0_ST_STM32_WINDOW_WATCHDOG_CLOCK_BITS,
```
Please use same set of #define's in whole driver (few lines below, you're using `DT_WWDT_0_IRQ`)

erwango reviewed (changes requested):
Few comments, otherwise I think we're fine.
You might want to remove the 'WIP' statement

dbkinder reviewed (approved):
+1 for doc changes

gon1332 commented:
Along with the others, with the latest force push, I've updated the `sample.yaml` of the watchdog sample for the sanitycheck to get the correct configurations.

erwango commented on drivers/watchdog/Kconfig.stm32:
```diff
@@ -30,3 +31,10 @@ config IWDG_STM32_TIMEOUT
 	  The min timeout supported is 0.1ms, the max timeout is 26214.4ms.
+
+config WWDG_STM32
+	bool "SYstem Window Watchdog (WWDG) Driver for STM32 family of MCUs"
```
s/SYstem/System

erwango commented on samples/drivers/watchdog/sample.yaml:
```diff
@@ -13,3 +13,32 @@ tests:
+    depends_on: watchdog
+    platform_whitelist: nucleo_f401re
+    extra_args: CONF_FILE="prj.conf;nucleo_f401re.conf"
+  sample.driver.window_watchdog_l496zg:
```
Is that the correct syntax? Should it be `window_watchdog_nucleo_l496zg`

gon1332 committed:
tests: drivers: watchdog: Update testcase

After the window watchdog introduction, the boards that support it over
independent watchdog fail to pass this test. This patch adds specific
tests for the nucleo_{f401re, l496zg} boards supporting WWDG.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 commented:
Rebased to master.

ioannisg commented:
@gon1332 @erwango do we still want this for 2.0 Release?
It needs, at least, a green CI for that

gon1332 commented:
@erwango This fails due to the last changes you've made on clock control for STM32 platform. The offending test is based on `nucleo_f401re` board. `nucleo_l496zg` is passing because the frequency needed is achieved with AHB prescaler of 1.

Should we throw the `nucleo_f401re` case for now?

ioannisg commented:
> 
> 
> @erwango This fails due to the last changes you've made on clock control for STM32 platform. The offending test is based on `nucleo_f401re` board. `nucleo_l496zg` is passing because the frequency needed is achieved with AHB prescaler of 1.
> 
> Should we throw the `nucleo_f401re` case for now?

Is there any chance we can sort this out without @erwango ?

gon1332 commented:
> > @erwango This fails due to the last changes you've made on clock control for STM32 platform. The offending test is based on `nucleo_f401re` board. `nucleo_l496zg` is passing because the frequency needed is achieved with AHB prescaler of 1.
> > Should we throw the `nucleo_f401re` case for now?
> 
> Is there any chance we can sort this out without @erwango ?

Yes. I'll do the what I've stated, and it'll be fine.

gon1332 commented:
@ioannisg Just force pushed. Hopefully these changes will green the CI.

ioannisg commented:
> 
> 
> @ioannisg Just force pushed. Hopefully these changes will green the CI.

Didn't,  so patch this is now pushed to next release.

gon1332 commented:
@erwango 
Updated PR to support all STM32 SOCs by including H7, G0, L1, WB and MP1.

gon1332 commented:
Rebased to HEAD.

arnop2 commented on dts/bindings/watchdog/st,stm32-window-watchdog.yaml:
```diff
@@ -0,0 +1,29 @@
+
+    interrupts:
+      category: required
+
```
extract from H7/MP1 user manual: " The early wakeup interrupt (EWI) can be used if specific safety operations or data logging must be performed before the actual reset is generated.
So should be "optional" instead of "required"

gon1332 replied:
Done

arnop2 commented on drivers/watchdog/wdt_wwdg_stm32.c:
```diff
@@ -0,0 +1,272 @@
+		LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_DBGMCU);
+#elif defined(CONFIG_SOC_SERIES_STM32L0X)
+		LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_DBGMCU);
+#endif
```
For some socs of the H7 series there are 2 WWDG instances with associated clocks 
WWDG1 on APB3 (Cortex-M7)
WWDG2 on APB1 (Cortex-M4)
Could be nice to support it also...

erwango replied:
@gon1332, if modifications for H7 are giving you headaches, you can remove for this PR. I asked support thinking this would be straight forward, since this is not the case, you can give up on this series, WWDG support will be added later. Sorry for trouble.

gon1332 replied:
The truth is that I don't have the hardware and time to study for these SOCs right now. If that's not a problem anyone could also collaborate on this PR. Otherwise, support for these SOCs could be added later.

erwango replied:
@gon1332, sure I understand, let add H7 later, otherwise is fine.

arnop2 replied:
seems reasonnable, ok for me
